phoenix-core/src/main/java/org/apache/phoenix/mapreduce/index/IndexUpgradeTool.java

```diff
@@ -0,0 +1,562 @@
+                    rebuildIndexes(dataTableFullName, conf);
+                }
+            } catch (IOException | SQLException | InterruptedException e) {
+                LOGGER.severe("Something went wrong while executing " + operation + " steps " + e);
+                return -1;
+            }
+        }
+        return 0;
+    }
+
+    private void modifyTable(Admin admin, String dataTableFullName, HashSet<String> indexes)
+            throws IOException {
+        if (upgrade) {
+            modifyIndexTable(admin, indexes);
+            modifyDataTable(admin, dataTableFullName);
+        } else {
+            modifyDataTable(admin, dataTableFullName);
+            modifyIndexTable(admin, indexes);
+        }
+    }
+
+    private void disableTable(Admin admin, String dataTable, HashSet<String>indexes)
+            throws IOException {
+        if (admin.isTableEnabled(TableName.valueOf(dataTable))) {
+            if (!dryRun) {
+                admin.disableTable(TableName.valueOf(dataTable));
+            }
+            LOGGER.info("Disabled data table " + dataTable);
+        } else {
+            LOGGER.info( "Data table " + dataTable +" is already disabled");
+        }
+        for (String indexName : indexes) {
+            if (admin.isTableEnabled(TableName.valueOf(indexName))) {
+                if (!dryRun) {
+                    admin.disableTable(TableName.valueOf(indexName));
+                }
+                LOGGER.info("Disabled index table " + indexName);
+            } else {
+                LOGGER.info( "Index table " + indexName +" is already disabled");
+            }
+        }
+    }
+
+    private void enableTable(Admin admin, String dataTable, HashSet<String>indexes)
+            throws IOException {
+        if (!admin.isTableEnabled(TableName.valueOf(dataTable))) {
+            if (!dryRun) {
+                admin.enableTable(TableName.valueOf(dataTable));
+            }
+            LOGGER.info("Enabled data table " + dataTable);
+        } else {
+            LOGGER.info( "Data table " + dataTable +" is already enabled");
+        }
+        for (String indexName : indexes) {
+            if(!admin.isTableEnabled(TableName.valueOf(indexName))) {
+                if (!dryRun) {
+                    admin.enableTable(TableName.valueOf(indexName));
+                }
+                LOGGER.info("Enabled index table " + indexName);
+            } else {
+                LOGGER.info( "Index table " + indexName +" is already enabled");
+            }
+        }
+    }
+
+    private void modifyDataTable(Admin admin, String tableName)
+            throws IOException {
+        TableDescriptorBuilder dataTableDescBuilder = TableDescriptorBuilder
+                .newBuilder(admin.getDescriptor(TableName.valueOf(tableName)));
+        if (upgrade) {
+            if (admin.getDescriptor(TableName.valueOf(tableName)).hasCoprocessor(Indexer.class.getName())) {
+                if (!dryRun) {
```

Comment:
Could simplify with helper methods for addCoprocessor(Class coprocessor) and removeCoprocessor(Class coprocessor) and then just have two lines for upgrade and two lines for rollback.